src/nwb_conversion_tools/tools/hdmf.py

```diff
@@ -1,7 +1,76 @@
+"""Collection of modifications of HDMF functions that are to be tested/used on this repo until propagation upstream."""
 from typing import Tuple
 
 import numpy as np
-from hdmf.data_utils import GenericDataChunkIterator
+from hdmf.data_utils import GenericDataChunkIterator as HDMFGenericDataChunkIterator
+
+
+class GenericDataChunkIterator(HDMFGenericDataChunkIterator):
+    def _get_default_buffer_shape(self, buffer_gb: float = 1.0) -> Tuple[int]:
+        num_axes = len(self.maxshape)
+        chunk_bytes = np.prod(self.chunk_shape) * self.dtype.itemsize
+        assert buffer_gb > 0, f"buffer_gb ({buffer_gb}) must be greater than zero!"
+        assert (
+            buffer_gb >= chunk_bytes / 1e9
+        ), f"buffer_gb ({buffer_gb}) must be greater than the chunk size ({chunk_bytes / 1e9})!"
+        assert all(
+            np.array(self.chunk_shape) > 0
+        ), f"Some dimensions of chunk_shape ({self.chunk_shape}) are less than zero!"
+
+        maxshape = np.array(self.maxshape)
+
+        # Early termination condition
+        if np.prod(maxshape) * self.dtype.itemsize / 1e9 < buffer_gb:
+            return tuple(self.maxshape)
+
+        buffer_bytes = chunk_bytes
+        axis_sizes_bytes = maxshape * self.dtype.itemsize
+        smallest_chunk_axis, second_smallest_chunk_axis, *_ = np.argsort(self.chunk_shape)
+        target_buffer_bytes = buffer_gb * 1e9
+
+        # If the smallest full axis does not fit within the buffer size, form a square along the two smallest axes
+        sub_square_buffer_shape = np.array(self.chunk_shape)
+        if min(axis_sizes_bytes) > target_buffer_bytes:
+            k1 = np.floor((target_buffer_bytes / chunk_bytes) ** 0.5)
+            for axis in [smallest_chunk_axis, second_smallest_chunk_axis]:
+                sub_square_buffer_shape[axis] = k1 * sub_square_buffer_shape[axis]
+            return tuple(sub_square_buffer_shape)
+
+        # Original one-shot estimation has good performance for certain shapes
+        chunk_to_buffer_ratio = buffer_gb * 1e9 / chunk_bytes
+        chunk_scaling_factor = np.floor(chunk_to_buffer_ratio ** (1 / num_axes))
+        unpadded_buffer_shape = [
```

Comment:
I meant that you could have vectorized the whole thing:
```python
unpadded_buffer_shape = np.clip(a=chunk_scaling_factor * np.array(self.chunk_shape), a_min=self.chunk_shape, a_max=self.maxshape).astype("int")
```

Reply:
Hm... I'll have to try that - maybe in a followup PR just so we can get this through